libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/ThunkActions.kt

```diff
@@ -0,0 +1,72 @@
+package com.mapbox.navigation.ui.app.internal
+
+import com.mapbox.navigation.ui.app.internal.destination.DestinationAction
+import com.mapbox.navigation.ui.app.internal.extension.ThunkAction
+import com.mapbox.navigation.ui.app.internal.navigation.NavigationState
+import com.mapbox.navigation.ui.app.internal.navigation.NavigationStateAction
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutesAction
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutesState
+import com.mapbox.navigation.utils.internal.ifNonNull
+import com.mapbox.navigation.utils.internal.toPoint
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.takeWhile
+import kotlinx.coroutines.launch
+
+/**
+ * End Navigation ThunkAction creator.
+ */
+fun endNavigation() = ThunkAction { store ->
+    store.dispatch(RoutesAction.SetRoutes(emptyList()))
+    store.dispatch(DestinationAction.SetDestination(null))
+    store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
+}
+
+/**
+ * Show Route Preview ThunkAction creator.
+ */
+fun CoroutineScope.showRoutePreview() = fetchRouteAndContinue { store ->
```

Comment:
I don't think this should be a global static. Many other architecture components include some level of scoping.

For example, why not put this in `NavigationViewApi` as a suspending function and then force the caller to handle the coroutine scope?

Otherwise, you could call this whenever and it'll be confusing when there is no reducer bound to consume the `NavigationStateAction`

Reply:
> I don't think this should be a global static. Many other architecture components include some level of scoping.

It's hidden in `com.mapbox.navigation.ui.app.internal` package, just like `Store`, `State` and `SharedApp`.

> For example, why not put this in NavigationViewApi as a suspending function and then force the caller to handle the coroutine scope?

Few reasons:
1. ThunkAction.invoke() is called by the `Store.dispatch()` method and is not a suspension function.
    ```kotlin
    fun interface ThunkAction {
        operator fun invoke(store: Store)
    }
    
    fun Store.dispatch(thunkAction: ThunkAction) {
        thunkAction(this) // <---shorthand for thunkAction.invoke(this)
    }
    ```
   The reason why it's implemented this way is to support various types of ThunkActions.
   ```kotlin
    fun thunkActionWithCallback() = ThunkAction { store -> 
        doSomeAsyncWork(callback = { result ->
            store.dispatch(DidFinishSomeWork(result))
        })
    }
    
    fun CoroutineScope.thunkActionWithCoroutine() = ThunkAction { store -> 
        launch {
            val result = doSomeMoreAsyncWork() // <-- calling suspension
            store.dispatch(DidFinishSomeWork(result))
        }
    }
    
    fun thunkActionThatDispatchesOtherActions() = ThunkAction { store -> 
        store.dispatch(Action1())
        store.dispatch(Action2())
    }
    
    fun main(scope: CoroutineScope, store: Store) {
        store.dispatch(thunkActionWithCallback())
        store.dispatch(thunkActionThatDispatchesOtherActions())
        store.dispatch(scope.thunkActionWithCoroutine())
    }
   ```

    ThunkAction creators that are suspension functions are not supported.
    ```kotlin
    suspend fun myActionCreator() = ThunkAction { store ->
        val result = doSomeWork() // <--- Suspension functions can be called only within the coroutine body.
        store.dispatch(DidFinishSomeWork(result))
    }
     
    suspend fun doSomeWork(): Result { ... }
    ```
2. Caller already handles the coroutine scope. The `showRoutePreview()` can only be called in the CoroutineScope owned by the caller.
3. Moving this to the `NavigtionViewApi` will require CorutineScope to be passed as an argument (unless we update to Kotlin 1.6.20 and use experimental [context receivers api](https://blog.jetbrains.com/kotlin/2022/04/kotlin-1-6-20-released/#prototype-of-context-receivers-for-kotlin-jvm) to provide CorutineScope).
4. Moving this to the `NavigtionViewApi` will require `NavigtionViewApi` to be injected into components that dispatch those actions.

Reply:
@tomaszrybakiewicz You already stated the reasons why we can't move this to `NavigationViewApi`, but it would be a bit weird for me as a customer to call half the public APIs via `navigationViewApi.enableRouteReplay` and the other half on the scope `scope.showRoutePreview()`. Some use cases that might be helpful:

#### Use case 1
User just uses Drop-In UI with no custom logic. In this case they did not invoke any APIs on `navigationViewAPI`

#### Use case 2
User want to add their own long click listener on the map -> fetch routes -> start route preview. In this case they would invoke the following public APIs
- fun fetchRoutes(options: RouteOptions) **or**
- fun fetchRoutes(points: List<Point>) **and**
- fun startRoutePreview()

**OR**
- fun startRoutePreview(options: RouteOptions) **or**
- fun startRoutePreview(points: List<Point>)

#### Use case 3
User is in free drive mode and directly wants to start navigation. In this case they would invoke the following public APIs.
- fun fetchRoutes(options: RouteOptions) **or**
- fun fetchRoutes(points: List<Point>) **and**
- fun setDestination(point: Point) **and**
- fun setRoutes(routes: List<NavigationRoute>) **and**
- fun startNavigation()

**OR**

- fun startNavigation(destination: Destination, routes: List<Routes>) **or**
- fun startNavigation(points: List<Point>) **or**
- fun startNavigation(routeOptions: RouteOptions)

#### Use case 4
User wants to end navigation
- fun stopNavigation()

With all these different use cases, it would be simple to make is so as a user I can just call `navigationViewApi` and I get a list of public APIs available as autocomplete. In case the user want to invoke `fetchRoutes`, it would be good to synchronously inform the user that they either have the route or it was a failure. Hence  changing
`fun fetchRoutes(options: RouteOptions)` in `main` to `suspend fun fetchRoutes(options: RouteOptions): SomeResult` would be a good improvement.

Reply:
I see the value in `ThunkActions` as it resolves this [ticket](https://github.com/mapbox/mapbox-navigation-android/issues/5953). But we should somehow figure out a way to expose the public APIs as mentioned above. We still have the chance to break `NavigationViewApi` apart if there are other solutions that would allow making the functions mentioned above public. 🤔 can we inject `coroutineScope` to `NavigationViewApi` as a constructor injection?

Reply:
> You already stated the reasons why we can't move this to NavigationViewApi, but it would be a bit weird for me as a customer to call half the public APIs via navigationViewApi.enableRouteReplay and the other half on the scope scope.showRoutePreview()

I think there is a bit of misunderstanding here. Introduced extension functions (`CoroutineScope.startActiveNavigation()`, `CoroutineScope.showRoutePreview()` and `endNavigation()`) are **action creators**. Those methods are internal and should never be called by a client code. Although their source visibility is public to allow access from libnavui-dropin module, they are hidden (`.internal` package). 
Furthermore, as I mentioned above, those methods are **action creators**. Calling each method will return a ThunkAction which needs to be dispatched (`store.dispatch(thunkAction)`) before its logic is executed.
So... calling the following has no effect.

```kotlin
// some Actvitiy
fun onResume() {
   ...
  lifecycleScope.startActiveNavigation()
  lifecycleScope.startActiveNavigation()
  lifecycleScope.showRoutePreview()
  lifecycleScope.showRoutePreview()
  endNavigation()
  endNavigation()
}
```